h2o-core/src/main/java/water/util/VecUtils.java

```diff
@@ -879,4 +882,93 @@ public void map(Chunk c, NewChunk nc) {
       result += mrt.result;
     }
   }
+
+  /**
+   * Randomly shuffle a Vec using Fisher Yates shuffle 
+   * https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+   */
+  public static class ShuffleVecTask extends MRTask<ShuffleVecTask> {
+    private final long _seed;
+    private final Vec _vec;
+    private transient int[] _localChunks;
+    private transient int[] _permutatedChunks;
+
+    public ShuffleVecTask(final Vec vec, final long seed) {
+      super();
+      _seed = seed;
+      _vec = vec;
+    }
+
+    @Override
+    protected void setupLocal() {
+      _localChunks = VecUtils.getLocalChunkIds(_vec);
+      _permutatedChunks = _localChunks.clone();
+      permute(_permutatedChunks, null);
+    }
+
+    private void permute(int[] arr, Random rng) {
+      if (null == rng) rng = getRNG(_seed);
+      for (int i = arr.length - 1; i > 0; i--) {
+        int j = rng.nextInt(i + 1);
+        final int old = arr[i];
+        arr[i] = arr[j];
+        arr[j] = old;
+      }
+    }
+
+    @Override
+    public void map(Chunk _cs, NewChunk nc) {
+      Random rng = getRNG(_seed + _cs.start());
+      Chunk cs = _vec.chunkForChunkIdx(_permutatedChunks[Arrays.binarySearch(_localChunks, _cs.cidx())]);
+
+      int[] permutedRows = IntStream.range(0, cs._len).toArray();
```

Comment:
probably doesn't matter but typically we don't use stream API inside MRTasks

Reply:
Thank you for the comment, I'll change that. Is there any technical reason for that? I can think of only being it too easy to do parallel stream operations within a task that should be single threaded.

Reply:
In this particular case, I wouldn't use it because it takes away the option to allocate the array using MemoryManager: Whenever you are allocating an array that is not small (= can grow arbitrarily) you should ideally use MemoryManager.

For small arrays located outside of the loop that processes the chunk items it doesn't matter, but it is not really the h2o-style.